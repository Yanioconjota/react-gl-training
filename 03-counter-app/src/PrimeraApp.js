import { Col } from "react-bootstrap";
import './PrimeraApp.scss';
import gif from './assets/img/Roque.gif';
import PropTypes from 'prop-types';
import CounterApp from "./CounterApp";

const PrimeraApp = ({ saludo, subtitulo }) => {
  const stringMsg = "Esto es un string!";
  const boolMsg = "Los booleans y los undefined no pueden ser impresos en el HTML";
  const arrMsg = ["🍲", "🍔", "🍟", "🍕", "🌭", "🥪"];
  const persona = { nombre: 'James', apellido: 'Howlett',  edad: '?' };
  console.log(saludo, subtitulo);

  /* if (!saludo) {
    throw new Error('El saludo es necesario');
  } */

  return (
    <Col className="mt-2">
      <h1 className="text-info">Impresión de variables en HTML</h1>
      <code>const stringMsg = 'this is a string!';</code>
      <p className="Code">{stringMsg}</p>
      <code>const boolMsg = true</code>
      <p className="Code">{boolMsg}</p>
      <code>const arrMsg = ["🍲", "🍔", "🍟", "🍕", "🌭", "🥪"]</code>
      <p className="Code">{arrMsg}</p>
      <code>
        const persona = {`{ nombre: 'James', apellido: 'Howlett',  edad: '?' }`}
      </code>
      <p className="Code">{`Error: Objects are not valid as a React child (found: object with keys {nombre, apellido, edad}). If you meant to render a collection of children, use an array instead.`}</p>
      <p className="Code">{`Se pueden usar string literals o JSON.stringify(persona, null, 3 para mostrar objeros:`}</p>
      <code className="Code">{JSON.stringify(persona, null, 3)}</code>
      <hr className="my-5" />
      <h1 className="text-info">Uso de props</h1>
      <code>
        const PrimeraApp ={" "}
        {`({ saludo = 'Hola insecto!'}) => {...
            ...export default PrimeraApp;`}
      </code>
      <p className="Code">
        {`({ saludo = 'Hola insecto!'}) => {...`} Es la desestructuración de las
        props o propiedades que recibe el componente, en este caso se le asigna
        un valor por defecto
      </p>
      <p className="Code">
        Se puede agregar una validación utilizando prop-types, obligando a
        asignar un tipo:
      </p>
      <code className="d-block Code">
        PrimeraApp.propTypes = {`{ saludo: PropTypes.string.isRequired }`}
      </code>
      <code className="d-block">{`<PrimeraApp saludo="Hola, Soy Goku!"/>`}</code>
      <code className="d-block Code">{`PrimeraApp.defaultProps = { subtitulo: 'Soy un subtítulo' }`}</code>
      <p className="Code">Asigna valores por defecto a las props</p>
      <p className="Code">{subtitulo}</p>
      <hr className="my-5" />
      <p>Tarea:</p>
      <ol>
        <li>
          Crear un nuevo componente dentro de la carpeta SRC llamado CounterApp
          ✔️
        </li>
        <li>El CounterApp debe de ser un Functional Component ✔️</li>
        <li>
          El contenido del CounterApp debe de ser:
          {`<h1>CounterApp</h1>
              <h2> {value} </h2>`}{" "}
          ✔️
        </li>
        <li>
          Donde "value" es una propiedad enviada desde el padre hacia el
          componente CounterApp (Debe ser númerica validada con PropTypes) ✔️
        </li>
        <li>
          Reemplazar o agregar en el index.js el componente de{" "}
          {`<PrimeraApp />`}
          por el componente {`<CounterApp />`} (no se olviden del value que debe
          de ser un número) ✔️
        </li>
        <li>
          Asegúrense de no tener errores ni warnings (Cualquier warning no
          usado, comentar el código) ✔️
        </li>
      </ol>
      <hr className="my-5" />
      <CounterApp
        value={666}
        title="Counter App tarea"
        msg="value desde el componente padre: "
        showButtons={false}
      />
      <CounterApp value={0} img={gif} />
      <hr />
      <h4 className="text-info">Introducción a las pruebas AAA:</h4>
      <ul>
        <li className="Code">Arrange (Arreglar)</li>
        <li className="Code">Act (Actuar)</li>
        <li className="Code">Assert (Afirmar)</li>
        <li>Primeras pruebas</li>
        <li>Jest</li>
        <li>Expect</li>
        <li>toBe</li>
        <li>Enzyme</li>
        <li>Comandos útiles en la terminal para pruebas</li>
        <li>Revisar elementos renderizados en el componente</li>
        <li>Simular eventos</li>
      </ul>
    </Col>
  );
};

PrimeraApp.propTypes = {
  saludo: PropTypes.string.isRequired
};

PrimeraApp.defaultProps = {
  subtitulo: 'Soy un subtítulo'
}

export default PrimeraApp;
