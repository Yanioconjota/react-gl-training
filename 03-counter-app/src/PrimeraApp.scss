@import './assets/scss/CustomBootstrap.scss';

body {
  width: 100%;
  color: white;
  background-color: #121113;
}

.PrimeraApp {
  text-align: center;
}

.App-logo {
  height: 35px;
  width: 35px;
  pointer-events: none;
  border-radius: 50%;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 5s linear;
  }
}

.Text-center {
  text-align: center;
}

hr {
  border: 1px dashed #61dafb;
  background-color: transparent;
}

#root {
  padding: 15px;
}

code {
  @extend .bg-dark;
  padding: 5px 12px;
  border-radius: 5px;
  margin: 5px 0;
}

.Code {
  @extend .text-yellow;
  margin: 5px 0;
  font-family: SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
}

.D-flex {
  display: flex;
  align-items: center;
  justify-content: flex-start;

  &.J-center {
    justify-content: center;
  }
}

.Title-flex {
  @extend .D-flex;
}

.Nav {
  @extend .D-flex;

  >* {
    color: #61dafb;
    text-decoration: none;
    margin: 5px;
  }
}