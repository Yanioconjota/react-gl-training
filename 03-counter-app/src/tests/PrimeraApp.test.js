//import { render } from '@testing-library/react';
import "@testing-library/jest-dom";
import { shallow } from 'enzyme';
import PrimeraApp from "../PrimeraApp";

describe('Pruebas en <PrimeraApp />', () => {
  /* test('debe de mostrar el mensaje Hola, Soy Goku!', () => {
    const saludo = 'Hola, Soy Goku!';
    const { getByText } = render(<PrimeraApp saludo={saludo}/>);
    expect(getByText(saludo)).toBeInTheDocument();
  }); */
  test("debe mostrar <PrimeraApp /> Correctamente ", () => {
    const saludo = "Hola, Soy Goku!";
    const wrapper = shallow(<PrimeraApp saludo={saludo}/>);
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de mostrar el subtítulo enviado por props ', () => {
    const saludo = "Hola, Soy Goku!";
    const subtitulo = "Soy un subtítulo";
    const wrapper = shallow(
      <PrimeraApp 
        saludo={saludo}
        subtitulo={subtitulo}/>
    );
    const textoParrafo = wrapper.find("p.sub-title").text();
    console.log(textoParrafo);
    expect(textoParrafo).toBe(subtitulo);
  });
  
  
});
