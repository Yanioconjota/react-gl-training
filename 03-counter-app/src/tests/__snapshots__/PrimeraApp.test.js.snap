// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <PrimeraApp /> debe mostrar <PrimeraApp /> Correctamente  1`] = `
<Col
  className="mt-2"
>
  <h1
    className="text-info"
  >
    Impresión de variables en HTML
  </h1>
  <code>
    &lt;PrimeraApp saludo="Hola, Soy Goku!"/&gt;
  </code>
  <p
    className="Code"
  >
    Hola, Soy Goku!
  </p>
  <code>
    const stringMsg = 'this is a string!';
  </code>
  <p
    className="Code"
  >
    Esto es un string!
  </p>
  <code>
    const boolMsg = true
  </code>
  <p
    className="Code"
  >
    Los booleans y los undefined no pueden ser impresos en el HTML
  </p>
  <code>
    const arrMsg = ["🍲", "🍔", "🍟", "🍕", "🌭", "🥪"]
  </code>
  <p
    className="Code"
  >
    🍲
    🍔
    🍟
    🍕
    🌭
    🥪
  </p>
  <code>
    const persona = 
    { nombre: 'James', apellido: 'Howlett',  edad: '?' }
  </code>
  <p
    className="Code"
  >
    Error: Objects are not valid as a React child (found: object with keys {nombre, apellido, edad}). If you meant to render a collection of children, use an array instead.
  </p>
  <p
    className="Code"
  >
    Se pueden usar string literals o JSON.stringify(persona, null, 3 para mostrar objeros:
  </p>
  <code
    className="Code"
  >
    {
   "nombre": "James",
   "apellido": "Howlett",
   "edad": "?"
}
  </code>
  <hr
    className="my-5"
  />
  <h1
    className="text-info"
  >
    Uso de props
  </h1>
  <code>
    const PrimeraApp =
     
    ({ saludo = 'Hola insecto!'}) =&gt; {...
            ...export default PrimeraApp;
  </code>
  <p
    className="Code"
  >
    ({ saludo = 'Hola insecto!'}) =&gt; {...
     Es la desestructuración de las props o propiedades que recibe el componente, en este caso se le asigna un valor por defecto
  </p>
  <p
    className="Code"
  >
    Se puede agregar una validación utilizando prop-types, obligando a asignar un tipo:
  </p>
  <code
    className="d-block Code"
  >
    PrimeraApp.propTypes = 
    { saludo: PropTypes.string.isRequired }
  </code>
  <code
    className="d-block"
  >
    &lt;PrimeraApp saludo="Hola, Soy Goku!"/&gt;
  </code>
  <code
    className="d-block Code"
  >
    PrimeraApp.defaultProps = { subtitulo: 'Soy un subtítulo' }
  </code>
  <p
    className="Code"
  >
    Asigna valores por defecto a las props
  </p>
  <p
    className="Code sub-title"
  >
    Soy un subtítulo
  </p>
  <hr
    className="my-5"
  />
  <p>
    Tarea:
  </p>
  <ol>
    <li>
      Crear un nuevo componente dentro de la carpeta SRC llamado CounterApp ✔️
    </li>
    <li>
      El CounterApp debe de ser un Functional Component ✔️
    </li>
    <li>
      El contenido del CounterApp debe de ser:
      &lt;h1&gt;CounterApp&lt;/h1&gt;
              &lt;h2&gt; {value} &lt;/h2&gt;
       
      ✔️
    </li>
    <li>
      Donde "value" es una propiedad enviada desde el padre hacia el componente CounterApp (Debe ser númerica validada con PropTypes) ✔️
    </li>
    <li>
      Reemplazar o agregar en el index.js el componente de
       
      &lt;PrimeraApp /&gt;
      por el componente 
      &lt;CounterApp /&gt;
       (no se olviden del value que debe de ser un número) ✔️
    </li>
    <li>
      Asegúrense de no tener errores ni warnings (Cualquier warning no usado, comentar el código) ✔️
    </li>
  </ol>
  <hr
    className="my-5"
  />
  <CounterApp
    msg="value desde el componente padre: "
    showButtons={false}
    title="Counter App tarea"
    value={666}
  />
  <CounterApp
    msg="La respuesta es: "
    showButtons={false}
    title="Counter App duplicado"
    value={42}
  />
  <CounterApp
    img="Roque.gif"
    msg=""
    showButtons={true}
    title="Counter App"
    value={0}
  />
  <hr />
  <h4
    className="text-info"
  >
    Introducción a las pruebas AAA:
  </h4>
  <ul>
    <li
      className="Code"
    >
      Arrange (Arreglar)
    </li>
    <li
      className="Code"
    >
      Act (Actuar)
    </li>
    <li
      className="Code"
    >
      Assert (Afirmar)
    </li>
    <li>
      Primeras pruebas
    </li>
    <li>
      Jest
    </li>
    <li>
      Expect
    </li>
    <li>
      toBe
    </li>
    <li>
      Enzyme
    </li>
    <li>
      Comandos útiles en la terminal para pruebas
    </li>
    <li>
      Revisar elementos renderizados en el componente
    </li>
    <li>
      Simular eventos
    </li>
  </ul>
  <hr />
  <h4
    className="text-info"
  >
    Iniciando pruebas con Enzyme 
    <em>
      React 17
    </em>
  </h4>
  <ol>
    <li>
      Crear el archivo setupTests.js dentro de src
    </li>
    <li>
      instalar enzyme: 
      <code
        className="d-block"
      >
        npm i --save-dev enzyme
      </code>
    </li>
    <li>
      Instalar el adaptador:
      <code
        className="d-block"
      >
        npm install --save-dev @wojtekmaj/enzyme-adapter-react-17
      </code>
    </li>
    <li>
      Instalar los snapshots:
      <code
        className="d-block"
      >
        npm install --save-dev enzyme-to-json
      </code>
    </li>
    <li>
      Configurar el setupTests.js:
      <code
        className="d-block"
      >
        <ul
          className="list-unstyled"
        >
          <li>
            import Enzyme from 'enzyme';
          </li>
          <li>
            import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
          </li>
          <li>
            import { createSerializer } from 'enzyme-to-json';
          </li>
          <li>
            <br />
          </li>
          <li>
            Enzyme.configure({ adapter: new Adapter() });
          </li>
          <li>
            expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));
          </li>
        </ul>
      </code>
    </li>
    <li>
      Bajar y subir la terminal
    </li>
  </ol>
  <h4
    className="text-info"
  >
    Para activar el intellisense de enzyme y jest
  </h4>
  <ol>
    <li>
      Instalar:
      <code
        className="d-block"
      >
        npm i @types/jest
      </code>
      <code
        className="d-block"
      >
        npm i -g @types/jest
      </code>
    </li>
    <li>
      Opcinalmente pueden agregar un archivo jsconfig.json a la raíz del proyecto si el paso anterior no funciona
      <code
        className="d-block"
      >
        {"typeAcquisition": { "include": ["jest"]}}
            
      </code>
    </li>
  </ol>
</Col>
`;
